src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js

```diff
@@ -2016,7 +2016,7 @@ describe('ReactDOMComponent', () => {
       expect(el.hasAttribute('whatever')).toBe(false);
 
       expectDev(console.error.calls.argsFor(0)[0]).toContain(
-        'Warning: Invalid prop `whatever` on <div> tag',
+        'Warning: Received `true` for non-boolean attribute `whatever`',
```

Comment:
Why does it warn about being non-boolean for unknown attributes? It seems like we don't know whether an unknown attribute is actually a boolean attribute or not, so this could lead to false positives.

Reply:
Maybe we can tweak the wording here. But it is intentional that behavior is the same for known and unknown attributes.

Reply:
In what way is the behavior the same? Known boolean attributes will not warn because of the `propertyInfo` checks. I'm not sure I understand why we assume unknown attributes being passed booleans are not boolean attributes.

Reply:
>Known boolean attributes will not warn because of the propertyInfo checks.

Yes, that's what I meant. The only ones for which we pass booleans through are the booleans we know. We don't pass booleans through for either known non-booleans or unknowns. This keeps whether a certain attribute is known or unknown unobservable. Without this property we can't hide the fact that, for example, `src` was cut from the whitelist.

Reply:
So how can a user correctly use a boolean attribute that we don't have whitelisted?

I don't think behavior changes in previously-known attributes implies means implementation details are leaked; it just means attributes behavior has changed. Which is why this is part of a major release. Why not instead utilize warnings in this major release that would allow us to treat any unknown attribute with boolean values as a boolean attribute in the next?

> Of course that means we can never delete booleans from the whitelist. But that seems like a fair tradeoff.

I don't agree that having to maintain a boolean whitelist forever is a fair tradeoff, it introduces the same issues we had with the whitelist in the first place, just with a smaller subset of attributes.

Reply:
>So how can a user correctly use a boolean attribute that we don't have whitelisted?

By casting it to a boolean.

>I don't think behavior changes in previously-known attributes implies implementation details are leaked; it just means attributes behavior has changed.

I don’t understand what you mean. Can you elaborate on what you propose instead? I think we might be talking past each other.

Reply:
> I think we might be talking past each other.

That's totally possible 😄 

> By casting it to a boolean.

I'm not sure what you mean, I'm talking about attributes that should already be accepting booleans. With this PR, if I try and render:

```jsx
<div bool-attr={true} />
```

It will warn and it will not set `bool-attr`. Assuming `bool-attr` is some valid boolean attribute that we just don't have whitelisted, this behavior is wrong. It _is_ a boolean attribute and it should be set/unset with `true`/`false`. Casting to strings, in this case, would also be wrong because `<div bool-attr="false" />` is not the behavior the user would want.

> I don’t understand what you mean. Can you elaborate on what you propose instead? I think we might be talking past each other.

So using `src` as an example, I'm assuming the problem we're worried about is that `<img src={true} />` would be a problem since it's no longer whitelisted and it would be incorrectly treated as a boolean attribute. I'm suggesting that we accept this change, but add some warnings when previously-known attributes are passed invalid values.

So if you did `<img src={true} />` it would warn with something like:


>The `src` attribute expects a string but received a boolean `true`. This will cause React to >treat it as a boolean attribute, which is likely a mistake. If you intend to use a stringified >boolean value, make sure to stringify it beforehand.

This might involve keeping a DEV-only whitelist if we just warned, or adding the whitelist back for just the next major to warn + maintain behavior, but IMO it would provide us with a better long-term solution for 17+

Reply:
>I'm not sure what you mean, I'm talking about attributes that should already be accepting booleans.

All of these are in the whitelist. Are they not? It turned out to be pretty small (<20 attributes), and since HTML is not adding more boolean attributes at a giant pace, it seems okay to keep that whitelist.

>I'm suggesting that we accept this change, but add some warnings when previously-known attributes are passed invalid values.

Can you write up a comment with “before” and “after” for each case? We’ve spent a few hours with @sebmarkbage talking about it yesterday and there are subtle things that are easy to miss. If you flesh out a table I can probably point out why we didn’t go with that proposal.

Reply:
That's fair, I figured there's likely some internal context that I'm missing. If everyone is OK with forever maintaining the whitelist then my proposal is moot 😄 As long as custom elements can correctly define arbitrary boolean attributes.

Reply:
```js
<div />; // false 
<div bool-attr="" />; // true
```

Reply:
I don't think custom elements should accept booleans coerced to strings neither.

Reply:
It's easy enough for static attribute values but once you have to start dynamically setting and unsetting boolean attributes you have to start switching between `""` and `null` and remember those map to `true` and `false`, which is unfortunate.

Reply:
That's when we add them to the whitelist.

There's no consistent pattern for what booleans should do (clear attribute vs "off" vs "false" vs "no"). So we'll always need some whitelist.

In theory we could make the default behavior be whatever has the most attributes following that rule. Do we know for sure which that is?

Reply:
In my mind the goal is to reduce that inconsistency, which we maintain by coercing booleans to `"true"` or `"on"`. In the long-term I think it makes sense to require users to be explicit about attributes that those enumerated attributes that expect boolean-ish values, and that's something we can add warnings for.

> In theory we could make the default behavior be whatever has the most attributes following that rule. Do we know for sure which that is?

Looking at [this attribute table in the spec](https://html.spec.whatwg.org/#attributes-3) it seems clear that there are far more boolean attributes than there are attributes that accept `"true"`/`"false"` or `"on"`/`"off"`